### PR TITLE
Consider keywords after matching identifiers

### PR DESCRIPTION
**Pull request by Palaolitico**
@tenderlove, here you have the PR.
I have little experience with Ruby, so any suggestion will be very welcome. Thanks.

### CODE REVIEW

File: lib/tinygql/lexer.rb [hunk 1]
```diff
@@ -40,6 +40,15 @@ class Lexer
 
     KW_RE = /#{Regexp.union(KEYWORDS.sort)}\b/
 
+    KW_FP3 = KEYWORDS.filter {
+      |w| w.length >= 4
```
Comment (tenderlove) [inline lib/tinygql/lexer.rb:44]: `|w|` should be on the same line as `{`

**Review by tenderlove — COMMENTED** [review]
Thanks for sending this.  The way I understand the speedup in this PR is that, if the leading byte is not an ident byte, we speculate it's an identifier (calling `@scan.skip(IDENTIFIER)`). The problem is that it may not be an identifier, so we still have to double check using `return tk if @scan.skip(KW_RE) == @len`.

I _think_ this optimization makes sense.

This PR eliminates the perfect hash opting to use a regular hash `KW_FP3` with integer keys.  I wonder if we could use this patch but also use the perfect hash with an array lookup.  It feels like we could either make this faster, or reduce the diff (or both).  The perfect hash only needs two bytes, but this needs 3, so it seems we could reduce some method calls.

**Comment by tenderlove** [comment]
I tried combining this with a perfect hash in 28117d062a2252f51a2f927ff54a15c2aeaf48bb, but this still seems slightly faster. I'm not sure why though. It seems I could extract fewer bytes from the string (thus reducing method calls), but my patch seemed to slow things down.

Benchmarks on this branch:

```
ruby -I lib bin/bench.rb
ruby 3.4.0dev (2024-10-11T19:34:57Z master 5f62522d5b) +PRISM [arm64-darwin24]
Warming up --------------------------------------
        kitchen-sink     1.133k i/100ms
           negotiate    32.000 i/100ms
Calculating -------------------------------------
        kitchen-sink     11.385k (± 0.5%) i/s   (87.83 μs/i) -     57.783k in   5.075498s
           negotiate    321.361 (± 0.3%) i/s    (3.11 ms/i) -      1.632k in   5.078419s
############################## ALLOCATIONS ##############################
        kitchen-sink       229
           negotiate      7365
```

With my patch

```
ruby -I lib bin/bench.rb
ruby 3.4.0dev (2024-10-11T19:34:57Z master 5f62522d5b) +PRISM [arm64-darwin24]
Warming up --------------------------------------
        kitchen-sink     1.088k i/100ms
           negotiate    31.000 i/100ms
Calculating -------------------------------------
        kitchen-sink     10.959k (± 0.4%) i/s   (91.25 μs/i) -     55.488k in   5.063207s
           negotiate    315.166 (± 0.3%) i/s    (3.17 ms/i) -      1.581k in   5.016475s
############################## ALLOCATIONS ##############################
        kitchen-sink       229
           negotiate      7365
```

I'd expect the array lookup to beat the hash lookup, but the key calculation is probably too slow on my patch.  I'll try to look in to it more later, but suffice it to say we should stick with this PR for now. 😄

File: lib/tinygql/lexer.rb (hunk 1 above)
Comment (Palaolitico) [reply lib/tinygql/lexer.rb]: Done. Thanks

**Comment by Palaolitico** [comment]
Aaron, I was also surprised of the good performance of `KW_FP3` as a normal map. Nevertheless, there seem to be some possible explanations.

If I understood it correctly, the goal of `KW_LUT` was to recognize the matched keyword without extracting it (to avoid the allocation) and without checking against a map (that it's slow).
The purpose of `KW_FP3` is a bit different. Its main goal is to avoid checking against `KW_RE` as much as possible because that regex is pretty expensive. To do that, we need to compute a value in a much bigger space than the `KW_LUT`. `KW_LUT` is an array of 32 positions, with 17 filled with keywords. The probability of a proper identifier to be recognized as a possible keyword by `KW_LUT` is 17/32. Therefore, more than half of the proper identifiers will be checked against `KW_RE`. On the other hand, the fingerprint for `KW_FP3` computes a value in the range [0,43264], therefore, the probability now is only 17/43264 (this math is approximate and optimistic, because neither the fingerprint nor the data are uniform).

Consider that, in your original code, all the access to `KW_LUT` are successful, but most of the access to `KW_FP3` are failures. The more failures we get for `KW_FP3`, the better.

I tried a larger fingerprint using 256 instead of 13. With 256, collisions on the first bytes are avoided. But the performance is a bit worse. I don't know why; maybe 256*256 is too big and yjit generates worse code. On the other side, for the benchmarking data, when scaling by 13, all the failures are caused by differences out of the first 4 bytes (maybe it is an overoptimization for the benchmarking data?).

I also considered to use a bitmap instead of a normal map. That would require around 5.5Kb for the bitmap and something like the perfect hash to find out the matched keyword. But I didn't find a bitmap in the Ruby std lib. I found some native implementations but I was afraid of the problem you describe in the post: too many jumps between C and Ruby. I also consider to make a pure Ruby bitmap, but I don't know enough Ruby to choose proper int sizes to avoid boxing. Maybe it can be implemented on a (binary) string.

**Comment by tenderlove** [comment]
> The purpose of `KW_FP3` is a bit different. Its main goal is to avoid checking against `KW_RE` as much as possible because that regex is pretty expensive. To do that, we need to compute a value in a much bigger space than the `KW_LUT`. `KW_LUT` is an array of 32 positions, with 17 filled with keywords. The probability of a proper identifier to be recognized as a possible keyword by `KW_LUT` is 17/32. Therefore, more than half of the proper identifiers will be checked against `KW_RE`. On the other hand, the fingerprint for `KW_FP3` computes a value in the range [0,43264], therefore, the probability now is only 17/43264 (this math is approximate and optimistic, because neither the fingerprint nor the data are uniform

ahhh yes. This totally makes sense. Thank you for spelling it out. Great optimization, and thank you so much for the PR!

**Pull request merged**